src/core/ext/filters/client_channel/client_channel.c

```diff
@@ -1451,10 +1451,24 @@ static void on_external_watch_complete(grpc_exec_ctx *exec_ctx, void *arg,
 static void watch_connectivity_state_locked(grpc_exec_ctx *exec_ctx, void *arg,
                                             grpc_error *error_ignored) {
   external_connectivity_watcher *w = arg;
-  grpc_closure_init(&w->my_closure, on_external_watch_complete, w,
-                    grpc_schedule_on_exec_ctx);
-  grpc_connectivity_state_notify_on_state_change(
-      exec_ctx, &w->chand->state_tracker, w->state, &w->my_closure);
+  if (w->state != NULL) {
+    /* TODO: add to active list here */
+    grpc_closure_init(&w->my_closure, on_external_watch_complete, w,
+                      grpc_schedule_on_exec_ctx);
+    grpc_connectivity_state_notify_on_state_change(
+        exec_ctx, &w->chand->state_tracker, w->state, &w->my_closure);
+  } else {
+    /* TODO: find in active list */
+    external_connectivity_watcher *found = something;
+    /* TODO: if found: */
+    grpc_connectivity_state_notify_on_state_change(
+        exec_ctx, &w->chand->state_tracker, NULL, &found->my_closure);
+    grpc_pollset_set_del_pollset(exec_ctx, w->chand->interested_parties,
```

Comment:
btw running into tsan issues here. With no-ops in the "TODO"s here, the "concurrent connectivity" test fails gives multiple tsan errors, one of them being: `use of an invalid mutex` at this `grpc_pollset_set_del_pollset` call.

```
==================
WARNING: ThreadSanitizer: use of an invalid mutex (e.g. uninitialized or destroyed) (pid=53204)
    #0 pthread_mutex_lock /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_common_interceptors.inc:3600:3 (concurrent_connectivity_test+0x000000439bce)
    #1 gpr_mu_lock /usr/local/google/home/apolcyn/grpc/src/core/lib/support/sync_posix.c:60:3 (concurrent_connectivity_test+0x000000543a72)
    #2 pollset_set_del_pollset /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/ev_poll_posix.c:1212:3 (concurrent_connectivity_test+0x0000004f9ae2)
    #3 grpc_pollset_set_del_pollset /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/ev_posix.c:238:3 (concurrent_connectivity_test+0x0000004aabc7)
    #4 watch_connectivity_state_locked /usr/local/google/home/apolcyn/grpc/src/core/ext/filters/client_channel/client_channel.c:1477:5 (concurrent_connectivity_test+0x00000052a051)
    #5 grpc_combiner_continue_exec_ctx /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/combiner.c:325:5 (concurrent_connectivity_test+0x0000004ed588)
    #6 grpc_exec_ctx_flush /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/exec_ctx.c:83:17 (concurrent_connectivity_test+0x0000004ab67b)
    #7 closure_exec_thread_func /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/executor.c:93:7 (concurrent_connectivity_test+0x0000004ffcdf)
    #8 thread_body /usr/local/google/home/apolcyn/grpc/src/core/lib/support/thd_posix.c:57:3 (concurrent_connectivity_test+0x00000054451d)

  Mutex M238348 (0x7d4c0014fbd0) created at:
    #0 pthread_mutex_lock /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/tsan/../sanitizer_common/sanitizer_common_interceptors.inc:3600:3 (concurrent_connectivity_test+0x000000439bce)
    #1 gpr_mu_lock /usr/local/google/home/apolcyn/grpc/src/core/lib/support/sync_posix.c:60:3 (concurrent_connectivity_test+0x000000543a72)
    #2 pollset_set_del_pollset /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/ev_poll_posix.c:1212:3 (concurrent_connectivity_test+0x0000004f9ae2)
    #3 grpc_pollset_set_del_pollset /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/ev_posix.c:238:3 (concurrent_connectivity_test+0x0000004aabc7)
    #4 watch_connectivity_state_locked /usr/local/google/home/apolcyn/grpc/src/core/ext/filters/client_channel/client_channel.c:1477:5 (concurrent_connectivity_test+0x00000052a051)
    #5 grpc_combiner_continue_exec_ctx /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/combiner.c:325:5 (concurrent_connectivity_test+0x0000004ed588)
    #6 grpc_exec_ctx_flush /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/exec_ctx.c:83:17 (concurrent_connectivity_test+0x0000004ab67b)
    #7 closure_exec_thread_func /usr/local/google/home/apolcyn/grpc/src/core/lib/iomgr/executor.c:93:7 (concurrent_connectivity_test+0x0000004ffcdf)
    #8 thread_body /usr/local/google/home/apolcyn/grpc/src/core/lib/support/thd_posix.c:57:3 (concurrent_connectivity_test+0x00000054451d)

SUMMARY: ThreadSanitizer: use of an invalid mutex (e.g. uninitialized or destroyed) /usr/local/google/home/apolcyn/grpc/src/core/lib/support/sync_posix.c:60:3 in gpr_mu_lock
==================
FAILED: bins/tsan/concurrent_connectivity_test GRPC_POLL_STRATEGY=poll-cv [ret=66, pid=53204]
FAILED: bins/tsan/concurrent_connectivity_test GRPC_POLL_STRATEGY=epoll
FAILED: bins/tsan/concurrent_connectivity_test GRPC_POLL_STRATEGY=poll
FAILED: bins/tsan/concurrent_connectivity_test GRPC_POLL_STRATEGY=poll-cv
FAILED: Some tests failed
```

looking into this, but not sure if there's something obvious